archive/miu_language/basic.lean

```diff
@@ -0,0 +1,204 @@
+/-
+Copyright (c) 2020 Gihan Marasingha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gihan Marasingha
+-/
+import tactic.linarith
+
+/-!
+# An MIU Decision Procedure in Lean
+
+The [MIU formal system](https://en.wikipedia.org/wiki/MU_puzzle) was introduced by Douglas
+Hofstadter in the first chapter of his 1979 book,
+[Gödel, Escher, Bach](https://en.wikipedia.org/wiki/G%C3%B6del,_Escher,_Bach).
+The system is defined by four rules of inference, one axiom, and an alphabet of three symbols:
+`M`, `I`, and `U`.
+
+Hofstadter's central question is: can the string "MU" be derived?
+
+It transpires that there is a simple decision procedure for this system. A string is derivable if
+and only if it starts with `"M"`, contains no other `"M"`s, and the number of `"I"`s in the string
+is congruent to 1 or 2 modulo 3.
+
+## The MIU System
+
+An _atom_ is any one of `M`, `I` or `U`. A _string_ is a finite sequence of zero or more symbols.
+To simplify notation, we write a sequence `[I,U,U,M]`, for example, as `IUUM`.
+
+The four rules of inference are:
+
+1. xI → xIU,
+2. Mx → Mxx,
+3. xIIIy → xUy,
+4. xUUy → xy,
+
+where the notation α → β is to be interpreted as 'if α is derivable, then β is derivable'.
+
+Additionally, we have an axiom
+
+* `MI` is derivable.
```

Comment:
Presumably this is a distinction that Hofstadter makes, but it seems unnatural from an idiomatic point of view in a language with inductive types. More straightforward would be to say there are five constructors for an inductively defined predicate on strings. `MI` is only different in that it is a constructor without an argument.

I know this is a hassle, but my preference would be restructuring to be idiomatic, with a discussion of how this differs from Hofstadter's presentation.

Not as good, but perhaps still sufficient, would be to have an explicit discussion of this point explaining why this development deviates from that.

Reply:
I hope I've made cleared this up with my latest commit. Please do let me know before I resolve the conversation. I'll continue working on the comments for `decision_suf.lean`.